src/Build/Logging/ParallelLogger/ConsoleOutputAligner.cs

```diff
@@ -0,0 +1,143 @@
+using System.Text;
+using Microsoft.Build.Shared;
+
+namespace Microsoft.Build.BackEnd.Logging
+{
+    /// <summary>
+    /// Align output to multiple lines so no logged test is lost due to limited <see cref="Console.BufferWidth"/>.
+    /// During alignment optional prefix/indent is applied.
+    /// </summary>
+    /// <remarks>
+    /// This class is not thread safe.
+    /// </remarks>
+    internal class ConsoleOutputAligner
+    {
+        internal const int ConsoleTabWidth = 8;
+
+        private readonly StringBuilder _reusedStringBuilder = new(1024);
+        private readonly int _bufferWidth;
+        private readonly bool _alignMessages;
+
+        /// <summary>
+        /// Constructor.
+        /// </summary>
+        /// <param name="bufferWidth">Console buffer width. -1 if unknown/unlimited</param>
+        /// <param name="alignMessages">Whether messages are aligned/wrapped into console buffer width</param>
+        public ConsoleOutputAligner(int bufferWidth, bool alignMessages)
+        {
+            _bufferWidth = bufferWidth;
+            _alignMessages = alignMessages;
+        }
+
+        /// <summary>
+        /// Based on bufferWidth split message into multiple lines and indent if needed.
+        /// TAB character are interpreted by standard Console logic. 
+        /// </summary>
+        /// <param name="message">Input message. May contains tabs and new lines. Both \r\n and \n is supported but replaced into current environment new line.</param>
+        /// <param name="prefixAlreadyWritten">true if message already contains prefix (message context, timestamp, etc...).</param>
+        /// <param name="prefixWidth">Width of the prefix. Every line in result string will be indented by this number of spaces except 1st line with already written prefix.</param>
+        /// <returns>Aligned message ready to be written to Console</returns>
+        /// <remarks>
+        /// For optimization purposes this method uses single <see cref="StringBuilder"/> instance. This makes this method non thread safe.
+        /// Calling side is expected this usage is non-concurrent. This shall nto be an issue as it is expected that writing into Console shall be serialized anyway.
+        /// </remarks>
+        public string AlignConsoleOutput(string message, bool prefixAlreadyWritten, int prefixWidth)
+        {
+            int i = 0;
+            int j = message.IndexOfAny(MSBuildConstants.CrLf);
+
+            StringBuilder sb = _reusedStringBuilder;
+            // prepare reused StringBuilder instance for new use.
+            sb.Length = 0;
+            // The string contains new lines, treat each new line as a different string to format and send to the console
+            while (j >= 0)
+            {
+                AlignAndIndentLineOfMessage(sb, prefixAlreadyWritten, prefixWidth, message, i, j - i);
+                i = j + (message[j] == '\r' && (j + 1) < message.Length && message[j + 1] == '\n' ? 2 : 1);
```

Comment:
if I read right, if there is a `\r` with no following `\n`, this will treat it as a newline. (Not sure I see tests verifying this.) 

As far as I know there is no modern platform that treats this as a newline, so I wonder whether it would be better to ignore it entirely.

Reply:
I though about it. I was not sure what is correct reaction to \r but I decided to interpret it as new line with following reasoning:
Move to start of line and override whatever was there does not make sense in our event based logging approach. In context  of one `EventArg` why would user wants to overwrite something. In multiple messages it is also unusable as a) messages are delimited by `NewLine` and b) messages can be interleaved in concurrent build.
So using \r must be mistake in either user custom messages or logged substring of messages. Now, how can we sanitize this to user best benefit?

1. implement as console standard i.e. move cursor to indented start of line
    - will override already written text at this line and user can lost some info
    - will be inconsistent with file loggers
    - is most probably not user intent
2. ignore = replace by space
    - will be inconsistent with file logger file   
![image](https://user-images.githubusercontent.com/25249058/142421835-419f7889-5a5c-4a09-97e3-3f9e7fe98910.png)
    - is most probably not user intent
3. interpret as new line
   - correct interpretation of user mistake (missed \n, or mismatching \n <=> \r)
   - consistent with file logger
   - consistent with how most file viewers implements it (tried notepad, VS, VS Code)  
![image](https://user-images.githubusercontent.com/25249058/142424176-a2e63c3e-59b2-425d-a8e6-e23f445a5a4c.png)

I have picked 3 and still considering it as best option.

I will add unit test cases with explanation.

Reply:
seems reasonable

Reply:
Unit tests added